plutus-ir/src/Language/PlutusIR/Optimizer/DeadCode.hs

```diff
@@ -28,11 +29,12 @@ removeDeadBindings
     -> Term tyname name a
 removeDeadBindings t =
     let
+        tRen = PLC.runQuote $ PLC.rename t
```

Comment:
I'd prefer to just document that you need to have run this beforehand.

Reply:
I can start adding `newtype` wrappers in this PR (comments are completely unreliable), but I'm not sure about this, because then

1. A lot of testing code will need to manually rename things?
2. How often do you have a term that is already globally unique, but such that you haven't called `rename` over it one line above? Preserving global uniqueness is hard, we should only do this when absolutely necessary, it doesn't seem like well-separated PIR transformation passes will suffer from noticeable performance overhead if each of them starts renaming things before processing them. It is way more reliable to just satisfy the invariant rather than to require it to hold. Even when you impose an invariant via a `newtype` wrapper, it's still more reliable to fulfill it automatically, because someone might use the constructor of a data type (like we do with `NormalizeType` for example) or even `coerce` in case the constructor is not exported.

So I believe it's better to rename incoming things rather than outcoming things. This is kinda the essence of the new approach: if you need global uniqueness, then fulfill and maintain it locally, if you need to maintain it. At least I'd start with the local maintenance approach and move to the global approach as an optimization (provided we need this optimization, which I think is questionable).

Reply:
On reflection I agree. It's probably not too wasteful. In that case I think you can remove the two calls to `rename` bracketing `compile` at least.

Reply:
However, I still think the newtypes would be helpful!